File: suc/swift/052.swift
import SwiftUI

// MARK: - AppTokens
// 翻译 AppTokens 对象到 Swift 结构体，包含静态属性
struct AppTokens {
    struct Colors {
        static let primary = Color(hex: 0xFF111111)
        static let secondary = Color(hex: 0xFF444444)
        static let background = Color(hex: 0xFFF5F5F5)
        static let surface = Color(hex: 0xFFFFFFFF)
        static let outline = Color(hex: 0xFFCCCCCC)
        static let onPrimary = Color(hex: 0xFFFFFFFF)
        static let onSurface = Color(hex: 0xFF111111)
    }

    struct TypographyTokens {
        // Kotlin的dp/sp通常直接对应SwiftUI的pt
        static let display = Font.system(size: 28, weight: .bold)
        static let headline = Font.system(size: 18, weight: .semibold)
        static let body = Font.system(size: 14, weight: .regular)
    }

    struct Shapes {
        let radius: CGFloat
        var style: RoundedCornerStyle = .continuous // 默认使用连续圆角，视觉上更平滑

        static let small = Shapes(radius: 6)
        static let medium = Shapes(radius: 10)
        static let large = Shapes(radius: 14)
    }

    struct Spacing {
        static let sm: CGFloat = 8
        static let md: CGFloat = 12
        static let lg: CGFloat = 16
    }

    struct ShadowSpec {
        // Kotlin的ShadowSpec(elevation, radius, dy, opacity)
        // elevation在Compose中常与blur radius相关，这里Kotlin明确给出了radius。
        // SwiftUI的shadow(color:radius:x:y:)中，radius是模糊半径。
        // 我们将Kotlin的`radius`映射到SwiftUI的`blurRadius`，`dy`映射到`yOffset`。
        let blurRadius: CGFloat
        let yOffset: CGFloat
        let opacity: Double
    }
    struct ElevationMapping {
        // Kotlin: ShadowSpec(2.dp, 4.dp, 2.dp, 0.1f)
        // 对应 SwiftUI: blurRadius: 4, yOffset: 2, opacity: 0.1
        static let level1 = ShadowSpec(blurRadius: 4, yOffset: 2, opacity: 0.1)
    }
}

// MARK: - 辅助扩展
// 扩展 Color 以支持十六进制颜色初始化
extension Color {
    init(hex: UInt, alpha: Double = 1.0) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0,
            opacity: alpha
        )
    }
}

// 扩展 View 以支持使用 AppTokens.Shapes 定义的圆角
extension View {
    func cornerRadius(_ shape: AppTokens.Shapes) -> some View {
        self.cornerRadius(shape.radius)
    }
}

// MARK: - RootScreen
struct RootScreen: View {
    let tabs = ["All", "Enrolled", "Completed"]
    @State private var selectedTabIndex: Int = 0
    
    var body: some View {
        // ZStack 用于设置整个屏幕的背景色，并忽略安全区域
        ZStack {
            AppTokens.Colors.background.ignoresSafeArea(.all)

            VStack(spacing: AppTokens.Spacing.md) {
                Text("Course Catalog")
                    .font(AppTokens.TypographyTokens.display)
                    .foregroundColor(AppTokens.Colors.primary)
                    .frame(maxWidth: .infinity, alignment: .leading) // 文本左对齐

                // 自定义 TabRow 的等效实现
                CustomTabRow(tabs: tabs, selectedTabIndex: $selectedTabIndex)
                    .background(AppTokens.Colors.surface) // TabRow 的背景色
                    .cornerRadius(AppTokens.Shapes.small) // TabRow 容器的圆角

                // HorizontalPager 的等效实现
                // 使用 TabView 配合 .page 样式实现水平分页
                TabView(selection: $selectedTabIndex) {
                    ForEach(tabs.indices, id: \.self) { index in
                        // Group 用于在 ForEach 内部条件性渲染视图
                        Group {
                            switch index {
                            case 0:
                                CourseList(title: "All Courses", count: 6)
                            case 1:
                                CourseList(title: "Enrolled", count: 3)
                            case 2:
                                CourseList(title: "Completed", count: 4)
                            default:
                                EmptyView() // 默认情况，防止意外索引
                            }
                        }
                        .tag(index) // 必须设置 tag 以便 TabView 的 selection 绑定工作
                    }
                }
                .tabViewStyle(.page(indexDisplayMode: .never)) // 隐藏默认的分页指示器
                .animation(.easeInOut, value: selectedTabIndex) // 动画页面切换
            }
            .padding(.horizontal, AppTokens.Spacing.lg)
            .padding(.vertical, AppTokens.Spacing.md)
            .ignoresSafeArea(.keyboard, edges: .bottom) // 键盘弹出时自动调整布局
        }
    }
}

// MARK: - CustomTabRow
struct CustomTabRow: View {
    let tabs: [String]
    @Binding var selectedTabIndex: Int
    @Namespace private var tabNamespace // 用于匹配几何效果，实现下划线动画

    var body: some View {
        HStack(spacing: 0) { // 标签之间无间距，与 Compose TabRow 保持一致
            ForEach(tabs.indices, id: \.self) { index in
                Button(action: {
                    selectedTabIndex = index
                }) {
                    VStack(spacing: 0) { // 文本和下划线之间无间距
                        Text(tabs[index])
                            .font(selectedTabIndex == index ? AppTokens.TypographyTokens.headline : AppTokens.TypographyTokens.body)
                            .foregroundColor(selectedTabIndex == index ? AppTokens.Colors.primary : AppTokens.Colors.secondary)
                            .padding(.vertical, AppTokens.Spacing.md) // 标签文本的垂直内边距
                            .frame(maxWidth: .infinity) // 每个标签占据等宽空间
                        
                        // 下划线指示器
                        if selectedTabIndex == index {
                            Rectangle()
                                .fill(AppTokens.Colors.primary)
                                .frame(height: 2)
                                .matchedGeometryEffect(id: "tabUnderline", in: tabNamespace) // 动画效果
                        } else {
                            // 占位符，保持未选中标签的布局一致性
                            Rectangle()
                                .fill(Color.clear)
                                .frame(height: 2)
                        }
                    }
                }
                .buttonStyle(PlainButtonStyle()) // 移除按钮的默认样式（如蓝色高亮）
            }
        }
        .animation(.easeInOut(duration: 0.2), value: selectedTabIndex) // 动画下划线移动
    }
}


// MARK: - CourseList
struct CourseList: View {
    let title: String
    let count: Int

    var body: some View {
        ScrollView(.vertical, showsIndicators: false) { // 允许内容超出屏幕时滚动
            VStack(alignment: .leading, spacing: AppTokens.Spacing.md) {
                Text(title)
                    .font(AppTokens.TypographyTokens.headline)
                    .foregroundColor(AppTokens.Colors.secondary)

                ForEach(0..<count, id: \.self) { index in
                    CourseCard(courseNumber: index + 1)
                }
            }
            .frame(maxWidth: .infinity) // 确保 VStack 在 ScrollView 中占据全部宽度
        }
    }
}

// MARK: - CourseCard
struct CourseCard: View {
    let courseNumber: Int

    var body: some View {
        // Card 等效实现：VStack 结合背景、圆角和阴影修饰符
        VStack {
            HStack(alignment: .center) { // 垂直居中对齐
                VStack(alignment: .leading, spacing: AppTokens.Spacing.sm) {
                    Text("Course \(courseNumber)")
                        .font(AppTokens.TypographyTokens.headline)
                        .foregroundColor(AppTokens.Colors.primary)
                    // Kotlin 的 (5 + it) 对应 Swift 的 (5 + courseNumber - 1)，因为 courseNumber 是 1-indexed
                    Text("Duration: \(5 + courseNumber - 1) hrs")
                        .font(AppTokens.TypographyTokens.body)
                        .foregroundColor(AppTokens.Colors.secondary)
                }
                Spacer() // 占据剩余空间，实现左右两端对齐

                // Box (按钮样式) 等效实现：Text 结合背景和胶囊形状裁剪
                Text("View")
                    .font(AppTokens.TypographyTokens.body)
                    .foregroundColor(AppTokens.Colors.onPrimary)
                    .padding(.horizontal, 12) // 水平内边距 12.dp
                    .frame(height: 24) // 高度 24.dp
                    .background(AppTokens.Colors.primary)
                    .clipShape(Capsule()) // 圆形或胶囊形裁剪，模拟 CircleShape
            }
            .padding(AppTokens.Spacing.md) // 内部内容与卡片边缘的内边距
        }
        .background(AppTokens.Colors.surface) // 卡片的背景色
        .cornerRadius(AppTokens.Shapes.medium) // 卡片的圆角
        .shadow(color: Color.black.opacity(AppTokens.ElevationMapping.level1.opacity), // 标准阴影颜色（黑色带透明度）
                radius: AppTokens.ElevationMapping.level1.blurRadius, // 模糊半径
                x: 0, // X 轴偏移，Kotlin 的 ShadowSpec 未指定，默认为 0
                y: AppTokens.ElevationMapping.level1.yOffset) // Y 轴偏移
        .frame(maxWidth: .infinity) // 占据全部可用宽度
    }
}


// MARK: - App 入口点
// 这是 SwiftUI 应用程序的入口结构体
@main
struct CourseCatalogApp: App {
    // UIViewControllerRepresentable 用于控制状态栏的可见性。
    // 这是在 SwiftUI 中隐藏状态栏最可靠的方法，尤其是在 iOS 16+。
    private struct StatusBarHidingViewController: UIViewControllerRepresentable {
        func makeUIViewController(context: Context) -> UIViewController {
            let controller = UIViewController()
            // 将控制器视图的背景设置为透明，确保它不会遮挡内容
            controller.view.backgroundColor = .clear
            return controller
        }

        func updateUIViewController(_ uiViewController: UIViewController, context: Context) {
            // 对于这个简单的控制器，无需更新
        }

        // 此方法由系统调用以确定状态栏的可见性
        func prefersStatusBarHidden() -> Bool {
            return true
        }
    }

    var body: some Scene {
        WindowGroup {
            RootScreen()
                .statusBarHidden(true)
                // 使内容填充整个屏幕，包括通常为状态栏和 Home 指示器保留的区域
                .ignoresSafeArea(.all)
                // 覆盖一个零大小的、隐藏状态栏的 UIViewController，以确保状态栏被隐藏。
                // frame(width: 0, height: 0) 使其不可见，但仍是视图层次结构的一部分，
                // 从而使其 prefersStatusBarHidden 方法生效。
                .overlay(StatusBarHidingViewController().frame(width: 0, height: 0))
        }
    }
}

// MARK: - 预览
struct RootScreen_Previews: PreviewProvider {
    static var previews: some View {
        RootScreen()
    }
}

File: suc/kotlin/052.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Shapes
import androidx.compose.material3.Surface
import androidx.compose.material3.Tab
import androidx.compose.material3.TabRow
import androidx.compose.material3.Text
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import kotlinx.coroutines.launch

private const val NAME = "076*CourseCatalog*en"
private const val UI_TYPE = "Education"
private const val STYLE_THEME = "Monochrome"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFF111111)
        val secondary = Color(0xFF444444)
        val background = Color(0xFFF5F5F5)
        val surface = Color(0xFFFFFFFF)
        val outline = Color(0xFFCCCCCC)
        val onPrimary = Color(0xFFFFFFFF)
        val onSurface = Color(0xFF111111)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val headline = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.SemiBold)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(10.dp)
        val large = RoundedCornerShape(14.dp)
    }
    object Spacing {
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.1f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onSurface,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    headlineMedium = AppTokens.TypographyTokens.headline,
    bodyMedium = AppTokens.TypographyTokens.body
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

@Composable
fun RootScreen() {
    val tabs = listOf("All", "Enrolled", "Completed")
    val pagerState = rememberPagerState(pageCount = { tabs.size })
    val scope = rememberCoroutineScope()
    Scaffold(
        contentWindowInsets = androidx.compose.foundation.layout.WindowInsets(0),
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .padding(horizontal = AppTokens.Spacing.lg, vertical = AppTokens.Spacing.md),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
        ) {
            Text("Course Catalog", style = AppTokens.TypographyTokens.display, color = AppTokens.Colors.primary)
            TabRow(
                selectedTabIndex = pagerState.currentPage,
                containerColor = AppTokens.Colors.surface,
                contentColor = AppTokens.Colors.primary
            ) {
                tabs.forEachIndexed { index, title ->
                    val selected = pagerState.currentPage == index
                    Tab(
                        selected = selected,
                        onClick = { scope.launch { pagerState.animateScrollToPage(index) } },
                        text = {
                            Text(
                                title,
                                style = if (selected) AppTokens.TypographyTokens.headline else AppTokens.TypographyTokens.body,
                                color = if (selected) AppTokens.Colors.primary else AppTokens.Colors.secondary
                            )
                        }
                    )
                }
            }
            HorizontalPager(state = pagerState, modifier = Modifier.fillMaxWidth()) { page ->
                when (page) {
                    0 -> CourseList("All Courses", 6)
                    1 -> CourseList("Enrolled", 3)
                    2 -> CourseList("Completed", 4)
                }
            }
        }
    }
}

@Composable
fun CourseList(title: String, count: Int) {
    Column(verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)) {
        Text(title, style = AppTokens.TypographyTokens.headline, color = AppTokens.Colors.secondary)
        repeat(count) {
            Card(
                shape = AppTokens.Shapes.medium,
                colors = CardDefaults.cardColors(containerColor = AppTokens.Colors.surface),
                elevation = CardDefaults.cardElevation(AppTokens.ElevationMapping.level1.elevation),
                modifier = Modifier.fillMaxWidth()
            ) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(AppTokens.Spacing.md),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Column(verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm)) {
                        Text("Course ${it + 1}", style = AppTokens.TypographyTokens.headline, color = AppTokens.Colors.primary)
                        Text("Duration: ${(5 + it)} hrs", style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.secondary)
                    }
                    Box(
                        modifier = Modifier
                            .height(24.dp)
                            .background(AppTokens.Colors.primary, CircleShape)
                            .padding(horizontal = 12.dp),
                        contentAlignment = Alignment.Center
                    ) {
                        Text("View", style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.onPrimary)
                    }
                }
            }
        }
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFF5F5F5)
@Composable
fun PreviewRoot() {
    AppTheme {
        RootScreen()
    }
}
